share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java

```diff
@@ -0,0 +1,315 @@
+        throw new RuntimeException("not implemented");
+    }
+
+    /**
+     * Used by share-partition leaders to write share-partition state to a share coordinator.
+     * This is an inter-broker RPC authorized as a cluster action.
+     *
+     * @param request WriteShareGroupStateParameters
+     * @return A completable future of WriteShareGroupStateResult
+     */
+    public CompletableFuture<WriteShareGroupStateResult> writeState(WriteShareGroupStateParameters request) throws IllegalArgumentException {
+        validate(request);
+        GroupTopicPartitionData<PartitionStateBatchData> gtp = request.groupTopicPartitionData();
+        String groupId = gtp.groupId();
+
+        Map<Uuid, Map<Integer, CompletableFuture<WriteShareGroupStateResponse>>> futureMap = new HashMap<>();
+        List<PersisterStateManager.WriteStateHandler> handlers = new ArrayList<>();
+
+        gtp.topicsData().forEach(topicData -> {
+            topicData.partitions().forEach(partitionData -> {
+                CompletableFuture<WriteShareGroupStateResponse> future = futureMap
+                    .computeIfAbsent(topicData.topicId(), k -> new HashMap<>())
+                    .computeIfAbsent(partitionData.partition(), k -> new CompletableFuture<>());
+
+                handlers.add(
+                    stateManager.new WriteStateHandler(
+                        groupId,
+                        topicData.topicId(),
+                        partitionData.partition(),
+                        partitionData.stateEpoch(),
+                        partitionData.leaderEpoch(),
+                        partitionData.startOffset(),
+                        partitionData.stateBatches(),
+                        future, null)
+                );
+            });
+        });
+
+        for (PersisterStateManager.PersisterStateManagerHandler handler : handlers) {
+            stateManager.enqueue(handler);
+        }
+
+        CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(
+            handlers.stream()
+                .map(PersisterStateManager.WriteStateHandler::result)
+                .toArray(CompletableFuture[]::new));
+
+        return combinedFuture.thenApply(v -> {
+            List<TopicData<PartitionErrorData>> topicsData = futureMap.keySet().stream()
+                .map(topicId -> {
+                    List<PartitionErrorData> partitionErrData = futureMap.get(topicId).entrySet().stream()
+                        .map(partitionFuture -> {
+                            int partition = partitionFuture.getKey();
+                            CompletableFuture<WriteShareGroupStateResponse> future = partitionFuture.getValue();
+                            try {
+                                WriteShareGroupStateResponse partitionResponse = future.get();
+                                return partitionResponse.data().results().get(0).partitions().stream()
```

Comment:
So in the default state persister read/write call, the sender can send a request like which can have:

```
groupId,
[
  topic1:[part1, part2],
  topic2:[part1],
  topic3:[part5]
 .
 .
 .
]
```

However, we are not sending these requests as is - we explode them, into multiple requests:

r1: groupId, topic1, part1
r2: groupId, topic1, part2
r3: groupId, topic2, part1
...

each request is then encapsulated in a handler and then enqueued in the PersisterStateManager. Exploding them is necessary since the share coordinator is keyed on a single (groupId, topic1, part1) so the entire request as sent by the caller is not serviceable by a single coordinator node.

One we get the results back from the handlers (futures complete). We combine the fragments back to a response appropriate for the original large request.